src/typechecker/FStar.TypeChecker.Cfg.fst

```diff
@@ -1063,17 +1090,104 @@ let built_in_primitive_steps : prim_step_set =
                 Some body
               | _ -> None)
             (fun body -> body)
-            (fun _t body -> Some body)
+            (fun _us _t body -> Some body)
+    in
+    let array_ops =
+      let of_list_op =
+        let emb_typ t = ET_app(PC.immutable_array_t_lid |> Ident.string_of_lid, [t]) in
+        let un_lazy universes t l r =
+          S.mk_Tm_app
+            (S.mk_Tm_uinst (U.fvar_const PC.immutable_array_of_list_lid) universes)
+            [S.iarg t; S.as_arg l]
+            r
+        in
+        (  PC.immutable_array_of_list_lid, 2, 1,
+           mixed_binary_op
+              (fun (elt_t, _) -> Some elt_t)
+              (fun (l, q) ->
+                match arg_as_list FStar.Syntax.Embeddings.e_any (l, q) with
+                | Some lst -> Some (l, lst)
+                | _ -> None)
+              (fun r (universes, elt_t, (l, blob)) ->
```

Comment:
this is embeds the result of the primop back as a term:

The representation is just a Tm_lazy, where the blob holds the native `ImmutableArray.t term`

The kind records the type of the blob as IA.t "any"
and the interesting thing here is that the thunk represents the blob back as pure F* term `IA.of_list u#universes elt_t l`. This unreduced representation can be used in a context where the blob doesn't ,make sense, e.g., in the SMT encoding, we represent the blob computed by `of_list l` just as the unreduced term `of_list l`.